### PR TITLE
[gui] Make proxy icon from statusbar clickable

### PR DESCRIPTION
**Pull request by mess110**
Clicking on the proxy icon will open settings showing the network tab

https://github.com/bitcoin/bitcoin/pull/11491#issuecomment-336685303

### CODE REVIEW

File: src/qt/bitcoingui.cpp [hunk 1]
```diff
@@ -661,12 +662,12 @@ void BitcoinGUI::trayIconActivated(QSystemTrayIcon::ActivationReason reason)
...
-        return;
+    openOptionsDialogWithTabIndex(0);
+}
 
-    OptionsDialog dlg(this, enableWallet);
-    dlg.setModel(clientModel->getOptionsModel());
-    dlg.exec();
+void BitcoinGUI::proxyIconClicked()
+{
+    openOptionsDialogWithTabIndex(2);
```
Comment (jonasschnelli) [inline src/qt/bitcoingui.cpp]: nit: not sure if setting by the object name would make more sense.
Something like `setCurrentWidget(ui->tabNetwork)`.

**Comment by jonasschnelli** [comment]
utACK a73098955f119690d270f8d4c872445873c41cc5

**Comment by sipa** [comment]
Concept ACK

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (mess110) [reply src/qt/bitcoingui.cpp]: Would need to get `ui->tabNetwork` and `ui->tabMain` at `BitcoinGUI` level and in my opinion that is not worth it. The main concern is related to magic values or if the tabs change. Magic values are not a real concern imo because they are "labeled" by the method calling them (ex: `proxyIconClicked`). I don't think the order of tabs will change. Will it?

I tried something similar to this:

```
     OptionsDialog dlg(this, enableWallet);
-    dlg.setCurrentTabIndex(tabIndex);
+    if (tabIndex == 0) {
+      dlg.setTabMain();
+    } else {
+      dlg.setTabNetwork();
+    }
```

It adds 2 new methods to `OptionsDialog`, but still needed to pass `tabIndex` because I would need to make a methods which return `ui->tabNetwork` and `ui->tabMain`.

So I think I prefer the current implementation. Opinions?

**Comment by promag** [comment]
Concept ACK.

File: src/qt/bitcoingui.cpp (hunk 1 above)
Comment (promag) [reply src/qt/bitcoingui.cpp]: Could create an enum for tab indexes? So it would be something like
```cpp
void BitcoinGUI::openOptionsDialogWithTab(OptionsDialog::Tab tab)
```

Comment (mess110) [reply src/qt/bitcoingui.cpp]: Added the `enum`, makes the code more readable, thx

**Comment by jonasschnelli** [comment]
Sorry for nitpicking...and thanks, it's better, but we still have static index positions (that is what I'm trying to avoid) which violates MVC (need to change the code if we change the tab order).


After this (see diff), it would be independent:
```diff
diff --git a/src/qt/optionsdialog.cpp b/src/qt/optionsdialog.cpp
index 8c9fc0def..5aeb659e3 100644
--- a/src/qt/optionsdialog.cpp
+++ b/src/qt/optionsdialog.cpp
@@ -172,9 +172,11 @@ void OptionsDialog::setModel(OptionsModel *_model)
 
 void OptionsDialog::setCurrentTab(OptionsDialog::Tab tab)
 {
-    int index = tab;
-    if (ui->tabWidget->currentIndex() != index) {
-        ui->tabWidget->setCurrentIndex(index);
+    QWidget *tab_widget = NULL;
+    if (tab == OptionsDialog::Tab::TAB_NETWORK) tab_widget = ui->tabNetwork;
+    if (tab == OptionsDialog::Tab::TAB_MAIN) tab_widget = ui->tabMain;
+    if (tab_widget && ui->tabWidget->currentWidget() != tab_widget) {
+        ui->tabWidget->setCurrentWidget(tab_widget);
     }
 }
 
diff --git a/src/qt/optionsdialog.h b/src/qt/optionsdialog.h
index ca537d92c..99b2fa7a2 100644
--- a/src/qt/optionsdialog.h
+++ b/src/qt/optionsdialog.h
@@ -41,8 +41,8 @@ public:
     ~OptionsDialog();
 
     enum Tab {
-        TAB_MAIN = 0,
-        TAB_NETWORK = 2,
+        TAB_MAIN,
+        TAB_NETWORK,
     };
 
     void setModel(OptionsModel *model);
```

**Comment by promag** [comment]
Agree with @jonasschnelli suggestion. nit `s/NULL/nullptr` above.

**Comment by mess110** [comment]
@jonasschnelli with me at least, don't be sorry for nitpicking. My goal is to write quality code and your suggestion is indeed an improvement. And it helps me learn. Thanks for the suggestion (@promag as well).

File: src/qt/bitcoingui.cpp [hunk 2]
```diff
@@ -250,6 +250,7 @@ BitcoinGUI::BitcoinGUI(interfaces::Node& node, const PlatformStyle *_platformSty
     subscribeToCoreSignals();
 
     connect(connectionsControl, SIGNAL(clicked(QPoint)), this, SLOT(toggleNetworkActive()));
```
Comment (MarcoFalke) [inline src/qt/bitcoingui.cpp]: Mind to use the new connect syntax here?

```cpp
connect(connectionsControl, &GUIUtil::ClickableLabel::clicked, this, &BitcoinGUI::toggleNetworkActive);

Comment (promag) [reply src/qt/bitcoingui.cpp]: Must change type of connectionsControl too

Comment (promag) [reply src/qt/bitcoingui.cpp]: Either wait for #13529 merge or do like 
https://github.com/bitcoin/bitcoin/pull/13529/files#diff-827883cef269471106f7a18fd8199c47R102

File: src/qt/bitcoingui.cpp [hunk 3]
```diff
@@ -193,7 +193,8 @@ BitcoinGUI::BitcoinGUI(interfaces::Node& node, const PlatformStyle *_platformSty
     // Subscribe to notifications from core
     subscribeToCoreSignals();
 
-    connect(connectionsControl, SIGNAL(clicked(QPoint)), this, SLOT(toggleNetworkActive()));
+    connect(connectionsControl, &GUIUtil::ClickableLabel::clicked, this, &BitcoinGUI::toggleNetworkActive);
+    connect(labelProxyIcon, SIGNAL(clicked(QPoint)), this, SLOT(proxyIconClicked()));
```
Comment (MarcoFalke) [inline src/qt/bitcoingui.cpp]: Could use new syntax here as well?

Comment (mess110) [reply src/qt/bitcoingui.cpp]: ops, I changed the wrong one and assumed the other ones will be merged in #13529. Fixed now

@MarcoFalke @promag I had to add `GUIUtil` namespace in the header file, hope the other PR will merge easily.

Should I squash the commits?

Comment (mess110) [reply src/qt/bitcoingui.cpp]: squashed

File: src/qt/bitcoingui.cpp [hunk 4]
```diff
@@ -193,7 +193,8 @@ BitcoinGUI::BitcoinGUI(interfaces::Node& node, const PlatformStyle *_platformSty
     // Subscribe to notifications from core
     subscribeToCoreSignals();
 
-    connect(connectionsControl, SIGNAL(clicked(QPoint)), this, SLOT(toggleNetworkActive()));
+    connect(connectionsControl, &GUIUtil::ClickableLabel::clicked, this, &BitcoinGUI::toggleNetworkActive);
+    connect(labelProxyIcon, &GUIUtil::ClickableLabel::clicked, this, &BitcoinGUI::proxyIconClicked);
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: nit, I now have a preference for:
```cpp
connect(labelProxyIcon, &GUIUtil::ClickableLabel::clicked, [this] {
    openOptionsDialogWithTab(OptionsDialog::TAB_NETWORK);
});
```
for these kind of private slots.

File: src/qt/optionsdialog.h [hunk 5]
```diff
@@ -50,7 +56,7 @@ private Q_SLOTS:
     void on_openBitcoinConfButton_clicked();
     void on_okButton_clicked();
     void on_cancelButton_clicked();
-
```
Comment (promag) [inline src/qt/optionsdialog.h]: nit, unnecessary change.

**Review by promag — COMMENTED** [review]
utACK 1081f76.

File: src/qt/optionsdialog.cpp [hunk 6]
```diff
@@ -170,6 +170,16 @@ void OptionsDialog::setModel(OptionsModel *_model)
     connect(ui->thirdPartyTxUrls, SIGNAL(textChanged(const QString &)), this, SLOT(showRestartWarning()));
 }
 
+void OptionsDialog::setCurrentTab(OptionsDialog::Tab tab)
+{
+    QWidget *tab_widget = nullptr;
```
Comment (promag) [inline src/qt/optionsdialog.cpp]: Suggestion:
```cpp
if (tab == OptionsDialog::Tab::TAB_NETWORK) {
    ui->tabWidget->setCurrentWidget(ui->tabNetwork);
} else {
    ui->tabWidget->setCurrentWidget(ui->tabMain);
}
```

File: src/qt/bitcoingui.cpp [hunk 7]
```diff
@@ -764,6 +764,17 @@ void BitcoinGUI::updateHeadersSyncProgressLabel()
         progressBarLabel->setText(tr("Syncing Headers (%1%)...").arg(QString::number(100.0 / (headersTipHeight+estHeadersLeft)*headersTipHeight, 'f', 1)));
 }
 
+void BitcoinGUI::openOptionsDialogWithTab(OptionsDialog::Tab tab)
+{
+    if(!clientModel || !clientModel->getOptionsModel())
```
Comment (promag) [inline src/qt/bitcoingui.cpp]: nit, add space after `if`.

**Review by promag — COMMENTED** [review]
utACK 76082cf.

File: src/qt/bitcoingui.cpp (hunk 7 above)
Comment (mess110) [reply src/qt/bitcoingui.cpp]: added

File: src/qt/optionsdialog.h (hunk 5 above)
Comment (promag) [reply src/qt/optionsdialog.h]: Could you remove this change to avoid conflict with #13753.

Comment (mess110) [reply src/qt/optionsdialog.h]: I removed it, even though intentionally adding whitespaces (the ones I removed) caused a bit of sadness inside. Even if it was for a good cause.

There was another instance, I removed that as well.

At the end of the day, all trailing whitespaces will be removed which gives me a warm fuzzy feeling.

Thx for the reviews btw

**Comment by laanwj** [comment]
works-for-me ACK 6d5fcad576962e5950641f7e7b113a6ac6f397e5
